Infer whether a closure implements `Fn`, `FnMut`, etc based on what actions it takes

nikomatsakis opened the pull request:
Currently, we only infer the kind of a closure based on the expected type or explicit annotation. If neither applies, we currently report an error. This pull request changes that case to defer the decision until we are able to analyze the actions of the closure: closures which mutate their environment require `FnMut`, closures which move out of their environment require `FnOnce`.

This PR is not the end of the story:

- It does not remove the explicit annotations nor disregard them. The latter is the logical next step to removing them (we'll need a snapshot before we can do anything anyhow). Disregarding explicit annotations might expose more bugs since right now all closures in libstd/rustc use explicit annotations or the expected type, so this inference never kicks in.
- The interaction with instantiating type parameter fallbacks leaves something to be desired. This is mostly just saying that the algorithm from https://github.com/rust-lang/rfcs/pull/213 needs to be implemented, which is a separate bug. There are some semi-subtle interactions though because not knowing whether a closure is `Fn` vs `FnMut` prevents us from resolving obligations like `F : FnMut(...)`, which can in turn prevent unification of some type parameters, which might (in turn) lead to undesired fallback. We can improve this situation however -- even if we don't know whether (or just how) `F : FnMut(..)` holds or not for some closure type `F`, we can still perform unification since we *do* know the argument and return types. Once kind inference is done, we can complete the `F : FnMut(..)` analysis -- which might yield an error if (e.g.) the `F` moves out of its environment. 

r? @nick29581

rust-highfive commented:
r? @pcwalton

(rust_highfive has picked a reviewer for you, use r? to override)

nikomatsakis commented:
cc @aturon

nikomatsakis commented:
Oh, a note to the eventual reviewer -- the commits are mostly independent, but there's a bit of back-and-forth because I realized my initial strategy was overly complex. In particular, I envisioned the "deferred resolutions" introduced in 9a85a17 as a more general thing, but I realized that because we don't have to worry about recursive closures for kind inference, we can simplify everything if we specialize the deferred resolutions to calls, which is done in f9e46a8.

eddyb commented on src/librustc/metadata/common.rs:
```diff
@@ -225,10 +226,7 @@ pub struct LinkMeta {
-pub const tag_closure: uint = 0x96;
-pub const tag_closure_type: uint = 0x97;
-pub const tag_closure_kind: uint = 0x98;
+// GAP 0x94...0x98
```
Could we not have this?

nikomatsakis replied:
I'm not sure I follow. Not have a gap? Not comment on the gap? I'm not sure how to avoid the gap except for renumbering everything below, which seems kind of pointless? (Or is there something obvious I'm overlooking?)

eddyb replied:
Forget about it, I'll remove it (the gap) whenever I touch this code again.

eddyb commented on src/librustc/metadata/common.rs:
```diff
@@ -139,14 +139,15 @@ pub enum astencode_tag { // Reserves 0x40 -- 0x5f
     tag_table_moves_map = 0x52,
     tag_table_capture_map = 0x53,
-    tag_table_closures = 0x54,
+    tag_table_closure_tys = 0x54,
```
Minor nit, you could add the new tag here and avoid changing `last_astencode_tag` (I cleaned up some of this mess in #21744).

nikomatsakis replied:
I also don't get this -- how could I avoid changing `last_astencode_tag`? I am converting one existing tag into two, and there doesn't seem to be a gap for me to take advantage of...

eddyb replied:
I mean, add the second one after the first one, to avoid changing *the name* of the last tag in the list. The values don't really matter as long as they don't overlap with anything else.

eddyb commented on src/librustc/middle/astencode.rs:
```diff
@@ -647,30 +647,7 @@ impl<'tcx> tr for MethodOrigin<'tcx> {
-            }
-        }
-    }).unwrap()
+    kind.encode(ebml_w).unwrap();
```
\o/

eddyb commented on src/librustc/middle/astencode.rs:
```diff
@@ -1782,35 +1769,23 @@ impl<'a, 'tcx> rbml_decoder_decoder_helpers<'tcx> for reader::Decoder<'a> {
+    fn read_closure_kind<'b, 'c>(&mut self, _dcx: &DecodeContext<'b, 'c, 'tcx>)
+                                 -> ty::ClosureKind
+    {
+        Decodable::decode(self).ok().unwrap()
```
I wouldn't add these if they're just calling `Decodable`.

eddyb commented on src/librustc/middle/astencode.rs:
```diff
@@ -1937,11 +1912,17 @@ fn decode_side_tables(dcx: &DecodeContext,
+                    c::tag_table_closure_tys => {
+                        let closure_ty =
+                            val_dsr.read_closure_ty(dcx);
+                        dcx.tcx.closure_tys.borrow_mut().insert(ast_util::local_def(id),
```
Do these ever have non-local `DefId`'s?

nikomatsakis replied:
I don't think they can, actually. At least not today. (But maybe if we adopt something like [this old RFC](https://github.com/rust-lang/rfcs/pull/105).)

eddyb commented on src/librustc_typeck/check/closure.rs:
```diff
@@ -45,17 +45,10 @@ pub fn check_expr_closure<'a,'tcx>(fcx: &FnCtxt<'a,'tcx>,
             // that, otherwise we'll error, requesting an annotation.
```
This comment wasn't updated.

eddyb commented on src/librustc_typeck/check/mod.rs:
```diff
@@ -170,8 +172,23 @@ pub struct Inherited<'a, 'tcx: 'a> {
+    // we may not have decided yet whether `c` is a `Fn`, `FnMut`, or
+    // `FnOnce` closure. In that case, we defer full resolution of the
+    // call until upvar inference can kick in and make the
+    // decision. We keep these deferred resolutions sorted by the
```
I think "sorted" is a bit confusing here - maybe "grouped" would be better?

eddyb commented on src/test/run-pass/unboxed-closures-infer-recursive-fn.rs:
```diff
@@ -0,0 +1,47 @@
+// closure.  As far as I can tell, coding up a recursive closure
+// requires the good ol' [Y Combinator].
+//
+// [Y Combinator]: http://en.wikipedia.org/wiki/Fixed-point_combinator#Y_combinator
```
Isn't it https://news.ycombinator.com?

eddyb commented:
I took the liberty of finding a few nits - feel free to r=me unless you want a second review.

Any reason why this PR doesn't disregard explicit kinds? Maybe error if inference and the explicit kind mismatches.

nikomatsakis commented:
> Any reason why this PR doesn't disregard explicit kinds? Maybe error if inference and the explicit kind mismatches.

No, I just ran out of time and decided to post it as is. Maybe I'll try it locally and see what happens. But it doesn't make sense to error: I actually expect inference to get different results than the expectation on a fairly regular basis. For example, if the expected kind is `FnMut`, but the closure is `|x| x*2`, then the inference would yield `Fn` -- but that's perfectly legal.

eddyb commented:
I was actually talking about `|&mut: x| x*2`, where one could change to `|&: x| x*2` so it still compiles at stage0.
I am thinking that errors would potentially catch some bugs before they're fully removed, like a corner case that causes some form of mutation to be ignored, for example.

nikomatsakis commented:
@eddyb as an experiment, I tried removing all inference of expected kinds beyond what's implemented in this PR. I immediately hit some (minor) bugs around `move` closures, which I have fixed. After that, everything bootstraps, but I get some failures in run-pass tests due to the problem I described in bullet 2 of the header: "We can improve this situation however -- even if we don't know whether (or just how) F : FnMut(..) holds or not for some closure type F, we can still perform unification since we do know the argument and return types. Once kind inference is done, we can complete the F : FnMut(..) analysis -- which might yield an error if (e.g.) the F moves out of its environment. "

nikomatsakis commented:
So I'm personally inclined to land it as is and patch those other problems later.

nikomatsakis commented:
Rebased, cleaned up @eddyb's nits, and fixed the [easy](https://github.com/nikomatsakis/rust/commit/a9c3841a501691f36f98c47402d679885989637f) [bugs](https://github.com/nikomatsakis/rust/commit/870aea216bb6f6d6a238c03e186a01bff8820ac0) that were exposed by relying entirely on closure inference, but not the harder ones.

nikomatsakis commented:
@eddyb want to look over those last two comments?

nikomatsakis commented:
reassigning to eddyb since he did a review already

eddyb commented:
@bors r+ 870aea2

bors commented:
:hourglass: Testing commit 870aea2 with merge 0ab8d5d...

bors commented:
:sunny: Test successful - [auto-linux-32-nopt-t](http://buildbot.rust-lang.org/builders/auto-linux-32-nopt-t/builds/3386), [auto-linux-32-opt](http://buildbot.rust-lang.org/builders/auto-linux-32-opt/builds/3387), [auto-linux-64-nopt-t](http://buildbot.rust-lang.org/builders/auto-linux-64-nopt-t/builds/3382), [auto-linux-64-opt](http://buildbot.rust-lang.org/builders/auto-linux-64-opt/builds/3390), [auto-linux-64-x-android-t](http://buildbot.rust-lang.org/builders/auto-linux-64-x-android-t/builds/3366), [auto-mac-32-opt](http://buildbot.rust-lang.org/builders/auto-mac-32-opt/builds/3403), [auto-mac-64-nopt-t](http://buildbot.rust-lang.org/builders/auto-mac-64-nopt-t/builds/3397), [auto-mac-64-opt](http://buildbot.rust-lang.org/builders/auto-mac-64-opt/builds/3393), [auto-win-32-nopt-t](http://buildbot.rust-lang.org/builders/auto-win-32-nopt-t/builds/3054), [auto-win-32-opt](http://buildbot.rust-lang.org/builders/auto-win-32-opt/builds/3052), [auto-win-64-nopt-t](http://buildbot.rust-lang.org/builders/auto-win-64-nopt-t/builds/1883), [auto-win-64-opt](http://buildbot.rust-lang.org/builders/auto-win-64-opt/builds/1891)
